nixos/modules/services/web-apps/bookstack.nix

```
@@ -290,50 +357,47 @@ in {
       wantedBy = [ "multi-user.target" ];
       serviceConfig = {
         Type = "oneshot";
+        RemainAfterExit = true;
         User = user;
         WorkingDirectory = "${bookstack}";
+        RuntimeDirectory = "bookstack/cache";
+        RuntimeDirectoryMode = 0700;
       };
-      script = ''
+      path = [ pkgs.replace-secret ];
+      script =
+        let
+          isSecret = v: isAttrs v && v ? _secret && isString v._secret;
+          bookstackEnvVars = lib.generators.toKeyValue {
+            mkKeyValue = lib.flip lib.generators.mkKeyValueDefault "=" {
+              mkValueString = v: with builtins;
+                if isInt         v then toString v
+                else if isString v then v
+                else if true  == v then "true"
+                else if false == v then "false"
+                else if isSecret v then v._secret
+                else throw "unsupported type ${typeOf v}: ${(lib.generators.toPretty {}) v}";
+            };
+          };
+          secretPaths = lib.mapAttrsToList (_: v: v._secret) (lib.filterAttrs (_: isSecret) cfg.config);
+          mkSecretReplacement = file: ''
+            replace-secret '${file}' '${file}' "${cfg.dataDir}/.env"
```

```suggestion
            replace-secret ${escapeShellArgs [ file file "${cfg.dataDir}/.env" ]}
```
could the setup unit be replaced by a preStart script on the main unit and the systemd credentials mechanism for secrets? (see "Credentials" in the systemd.exce manpage for reference)

---

While perhaps possible, it would be a bit messy - the main unit is defined by an instance of a submodule in the `phpfpm` module.